src/ucp/core/ucp_ep.c

```diff
@@ -834,11 +851,50 @@ void ucp_ep_destroy_internal(ucp_ep_h ep)
     ucp_ep_delete(ep);
 }
 
+static void ucp_ep_check_lanes(ucp_ep_h ep)
+{
+#if UCS_ENABLE_ASSERT
+    uint8_t num_inprog       = ep->num_discard_inprog + ep->num_flush_next_cnt;
+    uint8_t num_failed_tl_ep = 0;
+    ucp_lane_index_t lane;
+
+
+    for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+        num_failed_tl_ep += ucp_is_uct_ep_failed(ep->uct_eps[lane]);
+    }
+
+    ucs_assert((num_failed_tl_ep == 0) ||
+               (ucp_ep_num_lanes(ep) == num_failed_tl_ep));
+
+    /* An attempt to destroy UCT lanes can be done several times inside
+     * ucp_ep_cleanup_lanes()/ucp_ep_discard_lanes(), i.e. discarding lanes
+     * and then cleanup lanes during UCP EP close - so, if it happens, UCT
+     * lane has to be already set to '&ucp_failed_tl_ep' and UCP EP's
+     * 'ref_cnt' can be >= 1, when UCP EP is going to be destroyed.
+     * Otherwise - UCT lane is set to a real UCT EP and it means that
+     * discarding of UCT lanes was not started yet, i.e. called from UCP EP
+     * destroy function and UCP EP's ref_cnt has to be ('number of worker
+     * flush operations started for the UCP EP' + 'number of discarded
+     * operations started for the UCP EP' + '1').
+     * If UCT lane is destroyed, but some operations are still in-progress,
+     * completions can be received for this UCT EP - this will lead to
+     * undefined behavior, because a real UCT EP was already destroyed */
+    if (num_failed_tl_ep != 0) {
+        ucs_assert((ep->ref_cnt == num_inprog) ||
```

Comment:
`/* ep reference count is the number of outstanding flush operations and discards, plus 1 if not destroyed yet */`